datanode/networkhistory/fsutil/fsutil.go

```diff
@@ -212,3 +212,51 @@ func UntarFile(source io.Reader, decompressedFilesDestination string) error {
 	}
 	return nil
 }
+
+// ReadNetworkHistorySegmentData takes a io.Reader reading from a network history segment .tar archive then
+//   - looks inside the .tar archive for historysnapshot.tar.gz
+//   - looks looks historysnapshot.tar.gz for a file called `historyFileName`
+//   - returns an io.Reader for reading that file
+//
+// (all without needing to write to any temporary files)
+func ReadNetworkHistorySegmentData(outerTarFileReader io.Reader, historyFileName string) (io.Reader, error) {
+	outerTarReader := tar.NewReader(outerTarFileReader)
+
+	// Find historysnapshot.tar.gz in the top level segment tar archive
+	for {
+		header, err := outerTarReader.Next()
+		if err == io.EOF {
+			return nil, fmt.Errorf("couldn't find snapshot data in history segment")
+		}
+		if err != nil {
+			return nil, fmt.Errorf("error reading history segment tar file: %w", err)
+		}
+		if strings.Contains(header.Name, "historysnapshot.tar.gz") {
+			break
+		}
+	}
+
+	// Decompress historysnapshot.tar.gz
+	innerTarGzReader, err := gzip.NewReader(outerTarReader)
+	if err != nil {
+		return nil, fmt.Errorf("error creating gzip reader for inner tar file: %w", err)
+	}
+	defer innerTarGzReader.Close()
+
+	// Read the history .tar file
+	innerTarFile := tar.NewReader(innerTarGzReader)
+
+	// Find the table dump file in the inner .tar file
+	for {
+		header, err := innerTarFile.Next()
+		if err == io.EOF {
+			return nil, fmt.Errorf("table file '%s' not found in segment", historyFileName)
+		}
+		if err != nil {
+			return nil, fmt.Errorf("error reading inner tar file: %w", err)
+		}
+		if filepath.Base(header.Name) == historyFileName {
```

Comment:
I'm thinking about that issue we saw on the snapshot-pipeline where arbitrary non-vega-schema table was being picked up and added to the network history (was a ticket ever raised for that?)

Knowing that, is us not validating `historyFileName` against a list of known table names potentially a security issue (albeit unlikely) in that we are exposing an API that _could_ let someone read arbitrary non-vega tables from an unsuspecting data-node?

I feel like we should either validate against a known set of tables first in this API, or do something so that we're 100% sure the network history won't accidently contain non-vega tables

Reply:
That's a good point.

Though, I think we're OK in some sense because this is looking at network history files that are semi-public on IPFS at this point anyway; it can't hurt to do a bit of validation; I'll make a whitelist.